BootloaderCorePkg/Stage2/Stage2.c

```diff
@@ -36,6 +36,61 @@ LoadComponentCallback (
   }
 }
 
+/**
+  Extend EPayload to TPM.
+
+  @param[in]  ContainerSig    Containetr Signature
+  @param[in]  ComponentName   Component Name
+
+  @retval RETURN_SUCCESS      Operation completed successfully.
+  @retval RETURN_NOT_FOUND    EPayload hash Not found
+**/
+RETURN_STATUS
+ExtendEPayload (
+  UINT32    ContainerSig,
+  UINT32    ComponentName
+  )
+{
+  UINT8                         HashAlg;
+  UINT16                        HashSize;
+  UINT32                        Src;
+  UINT32                        Length;
+  UINT8                         DigestHash[HASH_DIGEST_MAX];
+  COMPONENT_ENTRY               *ComponentEntry;
+  EFI_STATUS                     Status;
+
+  Status = RETURN_SUCCESS;
+  ComponentEntry = NULL;
+  LocateComponentEntry (ContainerSig, ComponentName, NULL, &ComponentEntry);
+  if (ComponentEntry != NULL)  {
+
+    if ((ComponentEntry->AuthType == AUTH_TYPE_SHA2_256) || (ComponentEntry->AuthType == AUTH_TYPE_SIG_RSA2048_SHA256)){
+      HashAlg = HASH_TYPE_SHA256;
+      HashSize = SHA256_DIGEST_SIZE;
+    } else if ((ComponentEntry->AuthType == AUTH_TYPE_SHA2_384) || (ComponentEntry->AuthType == AUTH_TYPE_SIG_RSA3072_SHA384)){
+      HashAlg = HASH_TYPE_SHA384;
+      HashSize = SHA384_DIGEST_SIZE;
+    }
+
+    if (HashAlg == GetCryptoHashAlg(PcdGet32(PcdMeasuredBootHashMask))){
+      if (ComponentEntry->HashSize == HashSize) {
+        TpmExtendHash (ComponentEntry->HashData, HashAlg);
+      } else {
+        DEBUG ((DEBUG_INFO, "EPAYLOAD hash does not exist !"));
+        return RETURN_NOT_FOUND;
+      }
+    } else {
+      Status = GetComponentInfo (ContainerSig, &Src, &Length);

```

Comment:
why  do you need this ?    EPayload is always a container.    So if LocateComponentEntry() returned NULL, it is a failure condition already.

Reply:
This case addresses if measured boot hash type differs from container hash type.
Eg: SM3 used for MB and RSA3K/SHA384 for signing process

Reply:
Please add this explanation in the code comment appropriately so that future readers 
will know what/why is happening.

Reply:
If so, we are recalc the hash using the info from GetComponentInfo.  This is not ideal.   The base will point to data in flash.  The process will be very slow.
During loading, it has been copied into memory already.  Might be better to do it at that time to avoid performance issue.  All TPM extension can occur at the same place where we do hash verification.  Any thoughts ?

Reply:
Agreed. That is important to avoid reading from flash for hash calculation.

Reply:
We are performing TPM operations for stage components(in stage1B and Stage2) immediately after LoadComponentWithCallback() in which hash verification happens. Yes, We could use the "memory Buffer addr" returned from LoadComponentWithCallback() in case HashCalculation need to be done.

Reply:
Yes, maybe we can enhance the callback to support this.  Then the flow will be simpler to just extend hash in the callback function.